File: 07/sol_part2.py
# itoombes, Advent of Code 2025
# Day 7

def parse_input():
    if False:
        f = open("dummyInput.txt", "r")
    else:
        f = open("input.txt", "r")
    equations = list()
    for line in f.readlines():
        line = line.split(":")
        result = int(line[0])
        values = line[1][1:].split(" ")
        for i in range(0, len(values)):
            values[i] = int(values[i])
        equations.append((result, values))
    return equations 

def is_possible(equation):
    print(f"Validate {equation}")
    # Iterate thru each possible operation comparison
    ops = [0] * (len(equation[1]) - 1)
    while True:
        currentTotal = equation[1][0]
        # Go through variant
        for i in range(1, len(equation[1])):
            if ops[i - 1] == 1:
                currentTotal += equation[1][i]
            else if ops[i - 1] == 2:
                currentTotal = int(str(currentTotal) + str(equation[1][i]))
            else:
                currentTotal *= equation[1][i]
            # Skip results when already over total
            if currentTotal > equation[0]:
                break
            if currentTotal == equation[0] and i == len(equation[1]) - 1:
                print(f"\tValid : {ops}")
                return True
            print(f"\tInvalid : {currentTotal}")

        # Iterate to next variant
        i = 0
        while True: 
            if ops[i] == 0:
                ops[i] = 1
                for j in range(0, i):
                    ops[j] = 0
                break
            i += 1
            # If here, have no values that will match 
            if i == len(ops):
                print("\tNo valid values")
                return False


def main():
    equations = parse_input()
    count = 0
    for equation in equations:
        if is_possible(equation):
            count += equation[0]
            print(f"\t{count}")
    print(count)

if __name__ == "__main__":
    main()

File: 17/sol_part2.c
// itoombes, Advent of Code
// Day 17 Part 2

#include <stdio.h>

int program[] = {2, 4, 1, 1, 7, 5, 4, 0, 0, 3, 1, 6, 5, 5, 3, 0};

int get_print_value(int a) {
    return (((((a & 7) ^ 1) ^ (a >> ((a & 7) ^ 1))) ^ 6) & 7);
}

int main() {
    int a0 = 0;
    while (1) {
        int i = 0;
        int ai = a0;
        int valid = 1;
        while ((i < 3) && (valid == 1)) {
            if (get_print_value(ai) != program[i]) {
                valid = 0;
                break;
            }
            ai = ai >> 3;
            i++;
        }
        if ((valid == 1) && (get_print_value(ai) == program[i]) && (ai >> 3 == 0)) {
            break;
        }
        a0++;
    }

    printf("Found! %d", a0);

    while (a0 != 0) {
        printf("%d,", get_print_value(a0));
        a0 = a0 >> 3;
    }

    return 0;
}
